src/Symfony/Component/Form/Extension/Core/Type/DateTimeType.php

```diff
@@ -87,19 +90,24 @@ public function buildForm(FormBuilderInterface $builder, array $options)
         $timeFormat = self::DEFAULT_TIME_FORMAT;
         $calendar = \IntlDateFormatter::GREGORIAN;
         $pattern = is_string($options['format']) ? $options['format'] : null;
+        $immutable = 'datetimeimmutable' === $options['input'];
 
         if (!in_array($dateFormat, self::$acceptedFormats, true)) {
             throw new InvalidOptionsException('The "date_format" option must be one of the IntlDateFormatter constants (FULL, LONG, MEDIUM, SHORT) or a string representing a custom format.');
         }
 
         if ('single_text' === $options['widget']) {
             if (self::HTML5_FORMAT === $pattern) {
-                $builder->addViewTransformer(new DateTimeToRfc3339Transformer(
+                $viewTransformerClass = $immutable ? DateTimeImmutableToRfc3339Transformer::class : DateTimeToRfc3339Transformer::class;
```

Comment:
This is a good idea, although the constructors are already pretty lengthy, e.g. [`DateTimeToArrayTransformer::__construct()`](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Form/Extension/Core/DataTransformer/DateTimeToArrayTransformer.php).

I agree that the form Types probably know too much about the immutable handling than they should in their current form.